Fix invalid write in ds_queue_is_empty and mem leak while updating subjob counts

hirenvadalia opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
* Valgrind shows below leak and invalid write:
```
Invalid write of size 8
   at 0x47A2A4: ds_queue_is_empty (queue.c:181)
   by 0x4477FB: close_servers (pbs_sched.c:548)
   by 0x446CAD: die (pbs_sched.c:204)
   by 0x4486F7: main (pbs_sched.c:1044)
Address 0x8 is not stack'd, malloc'd or (recently) free'd

150 bytes in 1 blocks are possibly lost in loss record 1,008 of 1,446
   at 0x4C29133: malloc (vg_replace_malloc.c:299)
   by 0x465A4D: update_subjob_state_ct (array_func.c:578)
   by 0x4655C2: update_array_indices_remaining_attr (array_func.c:404)
   by 0x4657B9: chk_array_doneness (array_func.c:469)
   by 0x465985: update_subjob_state (array_func.c:520)
   by 0x4E8DA1: svr_setjobstate (svr_jobfunc.c:601)
   by 0x4CB316: post_sendmom (req_runjob.c:1613)
   by 0x54C6AF: dispatch_task (work_task.c:144)
   by 0x47C7CC: process_DreplyTPP (issue_request.c:875)
   by 0x49108F: is_request (node_manager.c:4280)
   by 0x49E09D: do_tpp (pbsd_main.c:395)
   by 0x49E139: tpp_request (pbsd_main.c:439)
```

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
* Added NULL check for given parameter in ds_queue_is_empty
* Modified update_subjob_state_ct() to use local buffer instead malloc-ed area
* Fixed typo in job state in on_job_exit

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->
* None

#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
[valgrind_after.txt](https://github.com/openpbs/openpbs/files/5340143/valgrind_after.txt)
[valgrind_before.txt](https://github.com/openpbs/openpbs/files/5340144/valgrind_before.txt)



<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

bhroam commented on src/server/array_func.c:
```diff
@@ -569,23 +569,20 @@ get_subjob_state(job *parent, int iindx)
-	sprintf(buf+strlen(buf), "%s:%d ", statename[JOB_STATE_EXITING],
-		pjob->ji_ajtrk->tkm_subjsct[JOB_STATE_EXITING]);
-	sprintf(buf+strlen(buf), "%s:%d ", statename[JOB_STATE_EXPIRED],
+	sprintf(buf, "%s:%d %s:%d %s:%d %s:%d",
```
Consider turning this into an snprintf().  It is unlikely to be necessary because BUF_SIZE is 512, and that is definitely large enough to store this mall string.

bhroam reviewed (approved):
Just one minor comment.  Just a suggestion.  I'll sign off.

vstumpf commented on src/server/array_func.c:
```diff
@@ -569,23 +569,20 @@ get_subjob_state(job *parent, int iindx)
 {
-	char *buf;
+	char buf[BUF_SIZE];
 	static char *statename[] = {
```
Hey, can you make this `const` as well? When it gets converted to c++, the compiler will complain.

bhroam commented on src/server/array_func.c:
```diff
@@ -570,12 +570,12 @@ void
 
 	buf[0] = '\0';
-	sprintf(buf, "%s:%d %s:%d %s:%d %s:%d",
+	snprintf(buf, sizeof(buf) - 1, "%s:%d %s:%d %s:%d %s:%d",
```
No need for the -1.  snprintf() handles that for you.

hirenvadalia commented:
@vstumpf I have addressed all comments and suggestions, If you are ok with changes can we please take this PR forward as there is one dependent work is waiting for these changes.

vstumpf reviewed (approved):
LGTM
